File: coding/src/com/algos/li15_tree/_Bt_L226_InvertBinaryTree.java
package  com.algos.li15_tree;

import  com.algos.li30_model.TreeNode;

public class _Bt_L226_InvertBinaryTree {

	public static void main(String[] args) {
		_Bt_L226_InvertBinaryTree obj = new _Bt_L226_InvertBinaryTree();
				
		TreeNode root = new TreeNode(1);
		root.left = new TreeNode(3);
		root.right = new TreeNode(4);
		root.left.left = new TreeNode(2);
        root.left.right = new TreeNode(6);
        root.left.right.left = new TreeNode(5);
		
		obj.invertTree(root);
	}
	
	// Problem: https://leetcode.com/problems/invert-binary-tree/
	// Idea: Recursively invert left and right sub treens and assign left to right and right left ptr's.
	// time: O(n)
    // space: O(1)
	public TreeNode invertTree(TreeNode root) {
		if(root ==  null) 
			return root;

		TreeNode left = invertTree(root.left);
		TreeNode right = invertTree(root.right);
		root.left = right;
		root.right = left;
        return root;
    }

}

File: dsa/topics/10_trees/bt/python/bt_l226__invert_binary_tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# Problem: https://leetcode.com/problems/invert-binary-tree/
# Idea: Recursively invert left and right subtrees and swap left and right pointers.
# Time: O(n)
# Space: O(1) ignoring recursion stack

def invert_tree(root):
    if not root:
        return root

    left = invert_tree(root.left)
    right = invert_tree(root.right)

    root.left = right
    root.right = left

    return root


# -------- Test --------
root = TreeNode(1)
root.left = TreeNode(3)
root.right = TreeNode(4)
root.left.left = TreeNode(2)
root.left.right = TreeNode(6)
root.left.right.left = TreeNode(5)

invert_tree(root)

# Function to do inorder traversal to check result
def inorder_traversal(node):
    if not node:
        return []
    return inorder_traversal(node.left) + [node.val] + inorder_traversal(node.right)

print(inorder_traversal(root))  # Expected inorder of inverted tree
